velero-plugins/imagestream/shared_test.go

```diff
@@ -0,0 +1,427 @@
+							ObjectStorage: &velerov1.ObjectStorageLocation{
+								Bucket: "buc",
+								Prefix: "prefix",
+							},
+						},
+						Credential: &corev1.SecretKeySelector{
+							LocalObjectReference: corev1.LocalObjectReference{
+								Name: "cloud-credentials",
+							},
+							Key: "cloud",
+						},
+					},
+				},
+			},
+			want:          []string{},
+			wantErr:       true,
+			wantErrString: "secrets \"oadp-bsl1-aws-registry-secret\" not found",
+		},
+	}
+	scheme := runtime.NewScheme()
+	corev1.AddToScheme(scheme)
+	metav1.AddMetaToScheme(scheme)
+	velerov1.AddToScheme(scheme)
+	var bslCRD apiextensionsv1.CustomResourceDefinition
+	err := yaml.Unmarshal([]byte(bslCRDyaml), &bslCRD)
+	if err != nil {
+		panic(err)
+	}
+	testEnv := &envtest.Environment{
+		Scheme: scheme,
+		CRDs: []*apiextensionsv1.CustomResourceDefinition{
+			&bslCRD,
+		},
+	}
+	cfg, err := testEnv.Start()
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer testEnv.Stop()
+	clients.SetInClusterConfig(cfg)
+	client, err := dynamic.NewForConfig(cfg)
+	if err != nil {
+		panic(err)
+	}
+	// corev1client, err := corev1c.NewForConfig(cfg)
+	discoveryClient, err := discovery.NewDiscoveryClientForConfig(cfg)
+	if err != nil {
+		panic(err)
+	}
+	// GR
+	groupResources, err := restmapper.GetAPIGroupResources(discoveryClient)
+	if err != nil {
+		panic(err)
+	}
+	// GVR
+	rm := restmapper.NewDiscoveryRESTMapper(groupResources)
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			// cleanup funcs
+			var cleanupFuncs []func()
+			// create objects defined in the test case
+			for _, obj := range tt.objs {
+				uobjMap, err := runtime.DefaultUnstructuredConverter.ToUnstructured(obj)
+				if err != nil {
+					panic(err)
+				}
+				uobj := unstructured.Unstructured{Object: uobjMap}
+
+				mapping, err := rm.RESTMapping(schema.GroupKind{Group: uobj.GroupVersionKind().Group, Kind: uobj.GroupVersionKind().Kind}, uobj.GroupVersionKind().Version)
+				if err != nil {
+					panic(err)
+				}
+				resourceClient := client.Resource(mapping.Resource).Namespace(uobj.GetNamespace())
+				_, err = resourceClient.Create(context.Background(), &uobj, metav1.CreateOptions{})
+				if err != nil {
+					if !apierrs.IsAlreadyExists(err) {
+						panic(err)
+					}
+				}
+				// don't cleanup namespace since its problematic with envtest https://github.com/kubernetes-sigs/controller-runtime/issues/880
+				if uobj.GroupVersionKind().Kind != "Namespace" {
+					cleanupFuncs = append(cleanupFuncs, func() {
+						resourceClient.Delete(context.Background(), uobj.GetName(), metav1.DeleteOptions{})
+					})
+				}
+			}
+			defer func() {
+				// cleanup after test
+				for i := range cleanupFuncs {
+					cleanupFuncs[i]()
+				}
+			}()
+			got, err := GetRegistryEnvsForLocation(tt.args.location, tt.args.namespace)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("GetRegistryEnvsForLocation() error = %v, wantErr %v", err, tt.wantErr)
+				return
```

Comment:
This is BAD

we never check error message like this

I suggest simplifying here, by removing `wantErr` parameter and checking if error is not nil and if `err.Error()` is not equal to expected error string

Reply:
These are how [generated tests](https://github.com/cweill/gotests/blob/16a93f6eb6519118b1d282e2f233596a98dd7e96/internal/render/templates/function.tmpl#L70) from [vscode-go](https://github.com/golang/vscode-go/blob/825de6bd32ae38bb4a6ae737e451f0ad98f1ecb2/extension/src/goGenerateTests.ts#L201) always is.. I simply added another err string check.

I can remove wantErr bool check to simplify.

Reply:
Many go communities have a pattern of checking bool for err != nil so I don't think this is an anti-pattern.

Reply:
Simplified in https://github.com/openshift/openshift-velero-plugin/compare/9c2af3f74b3901f0af3e7da3b36e7dd5a4ddf5a3..f7c1ec34924a8a977ca90c8f5c15305e6cf42df9